File: sample/src/main/java/de/blox/graphview/sample/GraphActivity.java
package de.blox.graphview.sample;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.android.material.snackbar.Snackbar;

import java.util.Objects;

import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.RecyclerView;
import de.blox.graphview.AbstractGraphAdapter;
import de.blox.graphview.Graph;
import de.blox.graphview.Node;

public abstract class GraphActivity extends AppCompatActivity {

    private int nodeCount = 1;

    private Node currentNode;

    protected RecyclerView graphView;
    protected AbstractGraphAdapter<NodeViewHolder> adapter;
    private FloatingActionButton fab;

    private void setupGraphView(Graph graph) {
        adapter = new AbstractGraphAdapter<NodeViewHolder>() {
            @NonNull
            @Override
            public NodeViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
                final View view = LayoutInflater.from(parent.getContext())
                        .inflate(R.layout.node, parent, false);
                return new NodeViewHolder(view);
            }

            @Override
            public void onBindViewHolder(@NonNull NodeViewHolder holder, int position) {
                holder.textView.setText(Objects.requireNonNull(getNodeData(position)).toString());
            }
        };

        graphView.setAdapter(adapter);
        adapter.submitGraph(graph);
    }

    protected abstract Graph createGraph();

    protected abstract void setLayoutManager();

    private void setupFab(final Graph graph) {
        fab = findViewById(R.id.addNode);
        fab.setOnClickListener(v -> {
            final Node newNode = new Node(getNodeText());

            if (currentNode != null) {
                graph.addEdge(currentNode, newNode);
            } else {
                graph.addNode(newNode);
            }
            adapter.notifyDataSetChanged();
        });

        fab.setOnLongClickListener(v -> {
            if (currentNode != null) {
                graph.removeNode(currentNode);
                currentNode = null;
                adapter.notifyDataSetChanged();
                fab.hide();
            }
            return true;
        });
    }

    private void setupToolbar() {
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        ActionBar ab = getSupportActionBar();
        if (ab != null) {
            ab.setHomeAsUpIndicator(R.drawable.ic_arrow_back);
            ab.setDisplayHomeAsUpEnabled(true);
        }
    }

    @Override
    public boolean onSupportNavigateUp() {
        onBackPressed();
        return true;
    }

    protected abstract void setEdgeDecoration();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_graph);

        final Graph graph = createGraph();
        setupToolbar();
        setupFab(graph);
        graphView = findViewById(R.id.graphView);
        setLayoutManager();
        setEdgeDecoration();
        setupGraphView(graph);

    }

    protected class NodeViewHolder extends RecyclerView.ViewHolder {

        TextView textView;

        NodeViewHolder(View itemView) {
            super(itemView);
            textView = itemView.findViewById(R.id.textView);

            itemView.setOnClickListener(v -> {
                if (!fab.isShown()) {
                    fab.show();
                }
                currentNode = adapter.getNode(getAdapterPosition());
                Snackbar.make(itemView, "Clicked on " + Objects.requireNonNull(adapter.getNodeData(getAdapterPosition())).toString(),
                        Snackbar.LENGTH_SHORT).show();
            });
        }
    }

    protected String getNodeText() {
        return "Node " + nodeCount++;
    }
}

File: sample/src/main/java/dev/bandb/graphview/sample/GraphActivity.kt
package dev.bandb.graphview.sample

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.google.android.material.snackbar.Snackbar
import dev.bandb.graphview.AbstractGraphAdapter
import dev.bandb.graphview.graph.Graph
import dev.bandb.graphview.graph.Node
import java.util.*

abstract class GraphActivity : AppCompatActivity() {
    protected lateinit var recyclerView: RecyclerView
    protected lateinit var adapter: AbstractGraphAdapter<NodeViewHolder>
    private lateinit var fab: FloatingActionButton
    private var currentNode: Node? = null
    private var nodeCount = 1

    protected abstract fun createGraph(): Graph
    protected abstract fun setLayoutManager()
    protected abstract fun setEdgeDecoration()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_graph)

        val graph = createGraph()
        recyclerView = findViewById(R.id.recycler)
        setLayoutManager()
        setEdgeDecoration()
        setupGraphView(graph)

        setupFab(graph)
        setupToolbar()
    }

    private fun setupGraphView(graph: Graph) {
        adapter = object : AbstractGraphAdapter<NodeViewHolder>() {
            override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): NodeViewHolder {
                val view = LayoutInflater.from(parent.context)
                        .inflate(R.layout.node, parent, false)
                return NodeViewHolder(view)
            }

            override fun onBindViewHolder(holder: NodeViewHolder, position: Int) {
                holder.textView.text = Objects.requireNonNull(getNodeData(position)).toString()
            }
        }.apply {
            this.submitGraph(graph)
            recyclerView.adapter = this
        }
    }

    private fun setupFab(graph: Graph) {
        fab = findViewById(R.id.addNode)
        fab.setOnClickListener {
            val newNode = Node(nodeText)
            if (currentNode != null) {
                graph.addEdge(currentNode!!, newNode)
            } else {
                graph.addNode(newNode)
            }
            adapter.notifyDataSetChanged()
        }
        fab.setOnLongClickListener {
            if (currentNode != null) {
                graph.removeNode(currentNode!!)
                currentNode = null
                adapter.notifyDataSetChanged()
                fab.hide()
            }
            true
        }
    }

    private fun setupToolbar() {
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        val ab = supportActionBar
        if (ab != null) {
            ab.setHomeAsUpIndicator(R.drawable.ic_arrow_back)
            ab.setDisplayHomeAsUpEnabled(true)
        }
    }

    override fun onSupportNavigateUp(): Boolean {
        onBackPressed()
        return true
    }

    protected inner class NodeViewHolder internal constructor(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var textView: TextView = itemView.findViewById(R.id.textView)

        init {
            itemView.setOnClickListener {
                if (!fab.isShown) {
                    fab.show()
                }
                currentNode = adapter.getNode(bindingAdapterPosition)
                Snackbar.make(itemView, "Clicked on " + adapter.getNodeData(bindingAdapterPosition)?.toString(),
                        Snackbar.LENGTH_SHORT).show()
            }
        }
    }

    protected val nodeText: String
        get() = "Node " + nodeCount++
}